File: src/main/java/controllers/products/ProductViewController.java
package controllers.products;

import javafx.application.Platform;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.scene.Node;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.control.cell.PropertyValueFactory;
import javafx.stage.Modality;
import javafx.stage.Stage;
import models.Product;
import models.ProductsData;
import models.treeView.SubcategoryItem;
import network.Api;
import network.RetrofitClient;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import utils.Dialogs;

import java.io.IOException;
import java.util.List;


public class ProductViewController {


    Api api = RetrofitClient.getApiService();

    private ObservableList<Product> productsOL = FXCollections.observableArrayList();
    private ObservableList<Product> productsFiltredOL = FXCollections.observableArrayList();
    private List<Product> productsList;
    private Product selectProduct;

    @FXML
    private TableView<Product> productTable;
    @FXML
    private TableColumn<Product, String> nameColumn;
    @FXML
    private TableColumn<Product, String> barcodeColumn;
    @FXML
    private TableColumn<Product, String> measurementColumn;
    @FXML
    private TableColumn<Product, String> countColumn;
    @FXML
    private TableColumn<Product, String> reserveColumn;
    @FXML
    private TableColumn<Product, String> priceColumn;
    @FXML
    private TreeView<SubcategoryItem> categoryTreeView;


    public ProductViewController() {

    }

    @FXML
    private void initialize() {
        // Инициализация таблицы адресатов с двумя столбцами.
        nameColumn.setCellValueFactory(new PropertyValueFactory<Product, String>("name"));
        measurementColumn.setCellValueFactory(new PropertyValueFactory<Product, String>("unit"));
        barcodeColumn.setCellValueFactory(new PropertyValueFactory<Product, String>("barcode"));
        countColumn.setCellValueFactory(new PropertyValueFactory<Product, String>("count"));
        reserveColumn.setCellValueFactory(new PropertyValueFactory<Product, String>("reserved"));
        priceColumn.setCellValueFactory(new PropertyValueFactory<Product, String>("priceFormat"));

        ContextMenu productTableContextMenu = new ContextMenu();
        MenuItem editProduct = new MenuItem("Редактировать");
        editProduct.setOnAction(new EventHandler<ActionEvent>() {
            @Override
            public void handle(ActionEvent event) {
                Product item = productTable.getSelectionModel().getSelectedItem();
                editProduct(item);
            }
        });

        MenuItem delProduct = new MenuItem("Удалить");
        delProduct.setOnAction(new EventHandler<ActionEvent>() {
            @Override
            public void handle(ActionEvent event) {
                Product item = productTable.getSelectionModel().getSelectedItem();
            }
        });
        MenuItem infoProduct = new MenuItem("Информация");
        infoProduct.setOnAction(new EventHandler<ActionEvent>() {
            @Override
            public void handle(ActionEvent event) {
                Product item = productTable.getSelectionModel().getSelectedItem();
            }
        });
        productTableContextMenu.getItems().setAll(editProduct,delProduct,infoProduct);
        productTable.setContextMenu(productTableContextMenu);


        loadProductsData();

        ContextMenu contextMenu = new ContextMenu();
        MenuItem edit = new MenuItem("Редактировать");
        edit.setOnAction(new EventHandler<ActionEvent>() {
            @Override
            public void handle(ActionEvent event) {
                SubcategoryItem item = categoryTreeView.getSelectionModel().getSelectedItem().getValue();
                if (item.getTrue_subcategory()){
                    //editSubcategory(Data.getSubcategoryById(item.getId(), subcategoryList));
                }else{
                    //editCategory(Data.getCategoryById(item.getId(), categoryList));
                }

            }
        });

        MenuItem del = new MenuItem("Удалить");
        del.setOnAction(new EventHandler<ActionEvent>() {
            @Override
            public void handle(ActionEvent event) {
                SubcategoryItem item = categoryTreeView.getSelectionModel().getSelectedItem().getValue();
                if (item.getTrue_subcategory()){

                }else{
                    //delCategory(Data.getCategoryById(item.getId(), categoryList));
                }

            }
        });
        contextMenu.getItems().setAll(edit, del);
        /*categoryTreeView.getSelectionModel().selectedItemProperty().addListener(
                (observable, oldValue, newValue) -> showProductsInCategory(newValue.getValue()));*/
        categoryTreeView.setContextMenu(contextMenu);

    }

    public void showAddProductDialog(ActionEvent actionEvent) {
        try {
            Stage stage = new Stage();
            FXMLLoader loader = new FXMLLoader();
            loader.setLocation(getClass().getResource("/view/products/ProductEditView.fxml"));
            Parent categoryAddFXML = loader.load();
            stage.setTitle("Новый товар");
            stage.setMinHeight(150);
            stage.setMinWidth(400);
            stage.setScene(new Scene(categoryAddFXML));
            stage.initModality(Modality.WINDOW_MODAL);
            stage.initOwner(((Node) actionEvent.getSource()).getScene().getWindow());

            ProductsEditController controller = loader.getController();
            //controller.setCategoriesAndSubcategories(categoryList, subcategoryList);

            stage.showAndWait();

            if (controller.isOk()) {
                loadProductsData();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void editProduct(Product product){
        try {
            Stage stage = new Stage();
            FXMLLoader loader = new FXMLLoader();
            loader.setLocation(getClass().getResource("/view/products/ProductEditView.fxml"));
            Parent categoryAddFXML = loader.load();
            stage.setTitle("Редактирование товара");
            stage.setMinHeight(150);
            stage.setMinWidth(400);
            stage.setScene(new Scene(categoryAddFXML));
            stage.initModality(Modality.WINDOW_MODAL);

            ProductsEditController controller = loader.getController();
            //controller.setCategoriesAndSubcategories(categoryList, subcategoryList);
            controller.setProduct(product);

            stage.showAndWait();

            if (controller.isOk()) {
                loadProductsData();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void delProduct(){

    }


    private void loadProductsData() {
        Call<ProductsData> call = api.productsData();
        call.enqueue(new Callback<ProductsData>() {
            @Override
            public void onResponse(Call<ProductsData> call, Response<ProductsData> response) {
                if (response.code() == 200) {
                    productsList = response.body().getProducts();
                    productsOL = FXCollections.observableArrayList(response.body().getProducts());
                    Platform.runLater(new Runnable() {
                        @Override
                        public void run() {
                            productTable.setItems(productsOL);
                            //categoryTreeView.setRoot(Data.categoryTreeViewRootItem(response.body().getCategories(), response.body().getSubcategories()));
                        }
                    });

                } else {
                    Dialogs.INSTANCE.showErrorDialog("Ошибка запроса на сервер!");
                }
            }

            @Override
            public void onFailure(Call<ProductsData> call, Throwable t) {
                Dialogs.INSTANCE.showExeptionDialog(t.getMessage());
            }
        });
    }


    public void selectHandle(){
        selectProduct = productTable.getSelectionModel().getSelectedItem();
        close();
    }

    public Product getSelectProduct(){
        return selectProduct;
    }

    private void close(){
        Stage stage = (Stage) productTable.getScene().getWindow();
        stage.close();
    }






}

File: src/main/java/controllers/products/ProductViewController.kt
package controllers.products

import controllers.LoadController
import javafx.collections.FXCollections
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.scene.control.*
import javafx.scene.control.Alert.AlertType
import javafx.scene.control.cell.PropertyValueFactory
import javafx.stage.Modality
import kotlinx.coroutines.experimental.javafx.JavaFx
import kotlinx.coroutines.experimental.launch
import models.Node
import models.Operation
import models.Product
import utils.CloseListener
import utils.Dialogs


class ProductViewController : LoadController<Boolean>() {

    private var operationOL = FXCollections.observableArrayList<Operation>()
    private var productList: List<Product>? = null

    private var productsOL = FXCollections.observableArrayList<Product>()
    private var productsFiltredOL = FXCollections.observableArrayList<Product>()
    private var productsList: List<Product>? = null
    lateinit var selectProduct: Product

    @FXML private lateinit var productTable: TableView<Product>
    @FXML private lateinit var nameColumn: TableColumn<Product, String>
    @FXML private lateinit var barcodeColumn: TableColumn<Product, String>
    @FXML private lateinit var measurementColumn: TableColumn<Product, String>
    @FXML private lateinit var countColumn: TableColumn<Product, String>
    @FXML private lateinit var reserveColumn: TableColumn<Product, String>
    @FXML private lateinit var priceColumn: TableColumn<Product, String>
    @FXML private lateinit var nodeTreeView: TreeView<Node>

    @FXML private fun initialize() {
        nameColumn.setCellValueFactory(PropertyValueFactory("name"))
        measurementColumn.setCellValueFactory(PropertyValueFactory("unit"))
        barcodeColumn.setCellValueFactory(PropertyValueFactory("barcode"))
        countColumn.setCellValueFactory(PropertyValueFactory("count"))
        reserveColumn.setCellValueFactory(PropertyValueFactory("reserved"))
        priceColumn.setCellValueFactory(PropertyValueFactory("priceFormat"))

        val productTableContextMenu = ContextMenu()
        val editProduct = MenuItem("Редактировать")
        editProduct.onAction = EventHandler {
            val item = productTable.selectionModel.selectedItem
            editProduct(item)
        }

        val delProduct = MenuItem("Удалить")
        delProduct.onAction = EventHandler {
            val item = productTable.selectionModel.selectedItem
            delProduct(item)
        }
        val infoProduct = MenuItem("Информация")
        infoProduct.onAction = EventHandler {
            val item = productTable.selectionModel.selectedItem
            infoProduct(item)
        }
        productTableContextMenu.items.setAll(editProduct, delProduct, infoProduct)
        productTable.contextMenu = productTableContextMenu

        val treeItemContextMenu = ContextMenu()
        val addNode = MenuItem("Добавить раздел")
        addNode.onAction = EventHandler {
            val item = nodeTreeView.selectionModel.selectedItem.value
            createNode(item)
        }
        val delNode = MenuItem("Удалть раздел")
        delNode.onAction = EventHandler {
            val item = nodeTreeView.selectionModel.selectedItem.value
            deleteNode(item)
        }
        val addProduct = MenuItem("Добавить товар")
        addProduct.onAction = EventHandler {
            val item = nodeTreeView.selectionModel.selectedItem.value
            addProduct(item)
        }

        treeItemContextMenu.items.setAll(addNode, delNode, addProduct)
        nodeTreeView.contextMenu = treeItemContextMenu

        loadProductsData()


    }

    private fun addProduct(node: Node) {
        ProductsEditController.show(node = node, owner = productTable as javafx.scene.Node) {result ->
            if (result) {
                loadProductsData()
            }
        }
    }

    private fun editProduct(product: Product) {
        ProductsEditController.show(product = product, owner = productTable as javafx.scene.Node) {result ->
        if (result) {
            loadProductsData()
        }
    }

    }

    private fun delProduct(product: Product) {

    }

    private fun infoProduct(product: Product) {

    }


    private fun createNode(parentNode: Node) {
        val dialog = TextInputDialog()
        dialog.title = "Новый раздел"
        dialog.headerText = "Добавление раздела"
        dialog.contentText = "Имя:"
        val result = dialog.showAndWait()

        result.ifPresent({ name -> addNode(name, parentNode) })
    }


    private fun addNode(name: String, parentNode: Node) = launch(JavaFx) {
        val node = Node()
        node.parent = parentNode.id
        node.name = name
        val result = api.addNode(node).await()
        if (result.isSuccessful()) {
            Dialogs.showDialog("Раздел добавлен!")
            loadProductsData()
        } else {
            Dialogs.showExeptionDialog(result.error)
        }
    }

    private fun deleteNode(node: Node) {
        val alert = Alert(AlertType.CONFIRMATION)
        alert.title = "Удаление раздела"
        alert.headerText = "Удалить раздел: ${node.name}?"

        val result = alert.showAndWait()
        if (result.get() === ButtonType.OK) {
            delNode(node)
        } else {

        }
    }

    private fun delNode(node: Node) = launch(JavaFx) {
        val id = node.id.toString()
        val result = api.delNode(id).await()
        if (result.isSuccessful()) {
            Dialogs.showDialog("Раздел удален!")
            loadProductsData()
        } else {
            Dialogs.showExeptionDialog(result.error)
        }
    }


    private fun loadProductsData() = launch(JavaFx) {
        val result = api.productsData().await()
        if (!result.isSuccessful()) return@launch
        productsOL = FXCollections.observableArrayList(result.notNullResult().products)
        productTable.setItems(productsOL)

        val items = result.notNullResult().nodes.map { TreeItem(it) }
        val roots = mutableListOf<TreeItem<Node>>()
        items.forEach {
            if (it.value.parent == null) {
                roots.add(it)
            } else {
                items[it.value.parent!! - 1].children.add(it)
            }
        }
        val root = TreeItem<Node>(Node("Все"))
        root.children.addAll(roots)
        nodeTreeView.root = root
    }


    fun selectHandle() {
        selectProduct = productTable.selectionModel.selectedItem
        close(true)
    }

    companion object {
        fun show(owner: javafx.scene.Node, callback: CloseListener<Boolean>) {
            LoadController.show(owner, callback,
                    path = "/view/products/ProductsView.fxml",
                    title = "Товары",
                    isResizable = false,
                    modality = Modality.WINDOW_MODAL)
        }
    }
}

